opnode/chain/source.go

```diff
@@ -0,0 +1,84 @@
+package chain
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimistic-specs/opnode/eth"
+	"github.com/ethereum-optimism/optimistic-specs/opnode/rollup"
+	"github.com/ethereum-optimism/optimistic-specs/opnode/rollup/derive"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+// ChainSource provides access to the L1 and L2 block graph
+type ChainSource interface {
+	L1BlockRefByNumber(ctx context.Context, number *big.Int) (eth.L1BlockRef, error) // TODO: Implement by Hash Version as well
+	L2BlockRefByNumber(ctx context.Context, number *big.Int, genesis *rollup.Genesis) (eth.L2BlockRef, error)
+	L2BlockRefByHash(ctx context.Context, l2Hash common.Hash, genesis *rollup.Genesis) (eth.L2BlockRef, error)
+}
+
+var WrongChainErr = errors.New("wrong chain")
+var TooDeepReorgErr = errors.New("reorg is too deep")
+
+// TODO: Make configurable
+var MaxReorgDepth = 500
+
+// L1ChainClient is the subset of methods that Chain needs to determine the L1 block graph
```

Comment:
Subset of methods of ... `ethclient.Client`  I assume? Same for L2ChainClient.

Reply:
yep